src/adapter/src/coord/sequencer/inner/peek.rs

```diff
@@ -934,12 +914,27 @@ impl Coordinator {
             Some(path) => {
                 // For everything else, return the plan for the stage identified
                 // by the corresponding path.
-                let row = trace
-                    .into_iter()
-                    .find(|entry| entry.path == path)
+
+                // A helper struct hat removes the first (and by assumption -
+                // only) plan that matches the given path from the collected
+                // trace.
+                let mut remove_plan = |path: &'static str| {
+                    let index = trace.iter().position(|entry| entry.path == path);
+                    index.map(|index| trace.remove(index))
+                };
+
+                // For certain stages we want to return the resulting fast path
+                // plan instead of the selected stage if it is present.
+                let plan = if stage.show_fast_path() && !config.no_fast_path {
+                    remove_plan(NamedPlan::FastPath.path()).or_else(|| remove_plan(path))
+                } else {
+                    remove_plan(path)
+                };
+
+                let row = plan
                     .map(|entry| Row::pack_slice(&[Datum::from(entry.plan.as_str())]))
                     .ok_or_else(|| {
```

Comment:
> Maybe this whole ok_or_else call could also be factored out into a new function, because it's the same for MVs and indexes, except for the stmt_kind.

This is already planned for my next PR (see the [explain: create an OptimizerTrace::into_rows helper](https://github.com/MaterializeInc/materialize/pull/24661/commits/964847e62f6b6e00448710d387addd4b67a9725a) and [explain: simplify Coordinator::<statement_type>_explain methods](https://github.com/MaterializeInc/materialize/pull/24661/commits/085fb3982353467e93d2eaf5c5e8dbeeb96f7c50) commits in #24661).